File: backend/app/api/routes/dependencies.py
from typing import Dict, List

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException

from ...core.auth import get_current_user
from ...core.cache import Cache
from ...core.circuit_breaker import circuit_breaker
from ...core.logging_config import get_logger
from ...domains.ml.dependency_analysis import DependencyAnalysisService

router = APIRouter(prefix="/dependencies", tags=["dependencies"])
logger = get_logger(__name__)
cache = Cache()
dependency_service = DependencyAnalysisService()


@router.post("/analyze", response_model=List[Dict])
@circuit_breaker
async def analyze_dependencies(
    work_items: List[Dict],
    background_tasks: BackgroundTasks,
    current_user=Depends(get_current_user),
) -> List[Dict]:
    """
    Analyze dependencies between work items.

    Args:
        work_items: List of work items to analyze
        background_tasks: FastAPI background tasks
        current_user: Current authenticated user

    Returns:
        List of dependencies with confidence scores
    """
    try:
        # Check cache first
        cache_key = f"dependencies:{','.join(str(item['id']) for item in work_items)}"
        cached_result = await cache.get(cache_key)
        if cached_result:
            return cached_result

        # Analyze dependencies
        dependencies = await dependency_service.analyze_dependencies(work_items)

        # Cache results in background
        background_tasks.add_task(cache.set, cache_key, dependencies)

        return dependencies

    except Exception as e:
        logger.error(f"Failed to analyze dependencies: {str(e)}")
        raise HTTPException(status_code=500, detail="Dependency analysis failed")


@router.post("/chain", response_model=List[Dict])
@circuit_breaker
async def suggest_dependency_chain(
    work_items: List[Dict], target_date: str, current_user=Depends(get_current_user)
) -> List[Dict]:
    """
    Suggest optimal work item ordering based on dependencies.

    Args:
        work_items: List of work items to order
        target_date: Target completion date
        current_user: Current authenticated user

    Returns:
        Ordered list of work items with explanations
    """
    try:
        # Analyze dependencies first
        dependencies = await dependency_service.analyze_dependencies(work_items)

        # Create optimal chain
        chain = create_dependency_chain(dependencies, target_date)

        return chain

    except Exception as e:
        logger.error(f"Failed to create dependency chain: {str(e)}")
        raise HTTPException(status_code=500, detail="Chain creation failed")


def create_dependency_chain(dependencies: List[Dict], target_date: str) -> List[Dict]:
    """Create optimal chain of work items."""
    # Implement chain creation logic here
    pass

File: frontend/src/services/dependencies.ts
import { WorkItem, Dependency } from '../types';
import { api } from './api';

/**
 * Analyze dependencies between work items
 * @param workItems List of work items to analyze
 * @returns List of detected dependencies with confidence scores
 */
export const analyzeDependencies = async (workItems: WorkItem[]): Promise<Dependency[]> => {
  try {
    const { data } = await api.post('/dependencies/analyze', { work_items: workItems });
    return data;
  } catch (error) {
    console.error('Failed to analyze dependencies:', error);
    throw error;
  }
};

/**
 * Get suggestions for optimal work item ordering
 * @param workItems Work items to order
 * @param targetDate Target completion date
 * @returns Ordered list of work items with explanations
 */
export const suggestDependencyChain = async (
  workItems: WorkItem[],
  targetDate: string
): Promise<WorkItem[]> => {
  try {
    const { data } = await api.post('/dependencies/chain', {
      work_items: workItems,
      target_date: targetDate,
    });
    return data;
  } catch (error) {
    console.error('Failed to suggest dependency chain:', error);
    throw error;
  }
};
